File: simple_text_editor/kevin/text_editor.go
package texteditor

import "fmt"

type Operation struct {
	op        int
	input_str string
}

type TextEditor struct {
	input    string
	op_stack []Operation
}

func (t *TextEditor) append(str string) {
	t.input += str
	t.op_stack = append(t.op_stack, Operation{op: 1, input_str: str})
}

func (t *TextEditor) delete(k int) {
	t.op_stack = append(t.op_stack, Operation{op: 2, input_str: t.input[len(t.input)-k:]})
	t.input = t.input[:len(t.input)-k]
}

func (t *TextEditor) print(k int) {
	fmt.Println(t.input[:len(t.input)-k])
}

func (t *TextEditor) undo() {
	if len(t.op_stack) == 0 {
		return
	}
	operation := t.op_stack[len(t.op_stack)-1]
	t.op_stack = t.op_stack[:len(t.op_stack)-1]
	if operation.op == 1 {
		t.delete(len(operation.input_str))
	} else {
		t.append(operation.input_str)
	}

	fmt.Println(t.op_stack)
}

File: simple_text_editor/kevin/texteditor.js
'use strict'

const Pair = require('./pair')

class TextEditor {
  constructor () {
    this.inputStack = []
    this.string = ''
  }
  createOperation (op, word) {
    return new Pair(op, word)
  }
  append (input) {
    if (typeof input === 'string' || input instanceof String) {
      this.inputStack.push(this.createOperation(1, input))
      this.string += input
    } else {
      console.log(`Invalid input type. Type of ${input} is not of type string...`)
      process.exit(1)
    }
  }

  delete (k) {
    if (Number.isInteger(k)) {
      const suffix = this.string.slice(-k)
      this.string = this.string.substr(0, this.string.length - k)
      this.inputStack.push(this.createOperation(2, suffix))
    } else {
      console.log(`Tried to delete a non integer value: ${k}`)
      process.exit(1)
    }
  }

  print (k) {
    if (Number.isInteger(k)) {
      const output = this.string.slice(0, k)
      console.log(output)
      return output
    } else {
      console.log(`Tried to print a non integer amount of chars: ${k}`)
      process.exit(1)
    }
  }

  undo () {
    if (this.inputStack.length === 0) {
      return ''
    }
    const pair = this.inputStack.pop()
    if (pair.operation === 1) {
      const wordLen = pair.text.length
      return this.delete(wordLen)
    } else if (pair.operation === 2) {
      return this.append(pair.text)
    }
  }
}

module.exports = TextEditor
